File: 33.cpp
// try to discard an interval if a[mid] is not equal to target
// to discard an interval first see what relation we have with the boundary element 
// and then compare target with a[mid] and boundary element and discard the appropriate interval
class Solution {
public:
    int search(vector<int>& a, int target) {
        int l = 0 , r = a.size() -1  ; 
        while(r >=  l){
            int mid = (r+l)/2 ; 
            if(a[mid] == target){
                return mid ; 
            }
            if(a[mid] >= a[l]){
                if(a[mid]>= target && a[l] <= target){
                    r = mid -1  ;
                }
                else{
                    l = mid + 1 ;
                }
            }
            else{
                if(a[mid] <= target && a[r] >= target){
                    l = mid+1 ; 
                }
                else{
                    r = mid -1 ;
                }
            }
        }
        return -1 ;
    }
};

File: python_solutions/33.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # Solution Overview
        # we can think of performing the normal binary search with little bit of modification
        # in normal binary serach we divide the interval into two parts
        # [min,mid] and [mid,max]
        # in our modified binary serach first we will check that which interval is in inscreasing order
        # let's assume without loss of generality that [min,mid] is in increasing order
        # now we will try to discard either this interval or [mid,max] interval based on in which interval we have our target ( which we can find easily by comparing the value in nums[min] and nums[mid] with the target)
        # for more details please see the solution
        min_idx = 0
        max_idx = len(nums)-1
        while min_idx <= max_idx:
            mid_idx = min_idx + (max_idx - min_idx)//2
            if nums[mid_idx] == target:
                return mid_idx
            if nums[mid_idx] < nums[max_idx]:
                if target > nums[mid_idx] and target <= nums[max_idx]:
                    min_idx = mid_idx + 1
                else:
                    max_idx = mid_idx - 1
            else:
                if target >= nums[min_idx] and target < nums[mid_idx]:
                    max_idx = mid_idx - 1
                else:
                    min_idx = mid_idx + 1
        return -1
        
